src/javaharness/java/arcs/api/NativeParticleImpl.java

```diff
@@ -25,14 +23,81 @@ public void setSpec(ParticleSpec spec) {
   }
 
   @Override
-  public void callSetHandles(Map<String, Handle> handles) {
+  public void setHandles(Map<String, Handle> handles) {
     // TODO: Should be async.
     // TODO: add startBusy & doneBusy support.
     // TODO: Add errors parameter.
-    this.setHandles(handles);
+    this.handles = handles;
```

Comment:
You might need/want to make a defensive copy here, perhaps similar elsewhere in the code, I haven't reviewed callsite usages to see what happens to the passed-in map subsequently.

Reply:
thanks for pointing it out, i need to pay more attention. this particular map is constructed by the pec specifically for this call, is it a good practice to make defensive copies just in case?

Reply:
I think it is more important if it's framework type of code that's likely to be used by many other devs who might not know that calling the method essentially involves transferring ownership. I think it's talked about in `Effective Java`, here's a StackExchange on it: https://stackoverflow.com/a/9743672

Arguably not making a defensive copy is a form of premature optimization -- it's always safer to make one. Java is supposed to do what it can to make copies of things cheaper-ish, via pools, views (for unmodifiable variants) and so forth.